examples/cvector-generator/vanilla_pca.hpp

```diff
@@ -0,0 +1,314 @@
+            float similarity = (float)((float*) similarity_tensor->data)[0];
+
+            // Eigenvector is the second last node in the graph
+            // struct ggml_tensor * eigenvector_tensor = gf->nodes[gf->n_nodes-2];
+            struct ggml_tensor * eigenvector_tensor = ggml_graph_node(gf,ggml_graph_n_nodes(gf)-2);
+            float * eigenvector_data = (float*) malloc(ggml_nbytes(eigenvector_tensor));
+            ggml_backend_tensor_get(eigenvector_tensor, eigenvector_data, 0, ggml_nbytes(eigenvector_tensor));
+            b = eigenvector_data;
+
+            // Eigenvalue computation is 1 operation before eigenvector computation
+            // struct ggml_tensor * eigenvalue_tensor = gf->nodes[gf->n_nodes-3];
+            struct ggml_tensor * eigenvalue_tensor = ggml_graph_node(gf, ggml_graph_n_nodes(gf)-3);
+            eigenvalue = (float)((float*) eigenvalue_tensor->data)[0];
+
+            // Check if the similarity is close enough to 1, if so we converged and should break
+            if(1 - similarity < pca_params.tolerance)
+                break;
+        }
+
+        // Free memory
+        ggml_gallocr_free(allocr);
+        ggml_free(ctx);
+    }
+
+    // Store result
+    result.principal_component->data = b;
+    result.explained_variance = eigenvalue;
+    return;
+}
+
+static void run_single_pca(struct pca_params &pca_params,
+             struct ggml_tensor * X,
+             struct pca_result &result
+             ) {
+
+    ggml_set_name(X, "input_tensor");
+
+    int m = X->ne[1]; // Number of features
+
+    // Step 1. Initialize GGML Backend
+    ggml_backend_t backend = NULL;
+    #ifdef GGML_USE_CUDA
+        fprintf(stderr, "%s: using CUDA backend\n", __func__);
+        backend = ggml_backend_cuda_init(0); // init device 0
+        if (!backend) { fprintf(stderr, "%s: ggml_backend_cuda_init() failed\n", __func__); }
+    #endif
+    // If there aren't GPU Backends fallback to CPU backend
+    if (!backend) { backend = ggml_backend_cpu_init(); }
+
+    // Compute the context size needed
+    size_t ctx_size = 2 * ggml_tensor_overhead();
+
+    // Step 2. Initialize GGML Context
+    struct ggml_init_params ctx_params {
+        ctx_size,  // mem_size
+        NULL,      // mem_buffer
+        true,      // no_alloc
+    };
+    struct ggml_context * ctx = ggml_init(ctx_params);
+
+    ggml_backend_buffer_t buffer = ggml_backend_alloc_ctx_tensors(ctx, backend);
+
+    // Step 3. Compute the data covariance matrix
+    struct ggml_tensor  * covariance = ggml_new_tensor_2d(ctx, GGML_TYPE_F32, m, m);
+    ggml_set_name(covariance, "covariance_tensor");
+    compute_covariance(pca_params, X, covariance, backend);
+
+    // Step 4. Power iteration
+    result.principal_component = ggml_new_tensor_1d(ctx, GGML_TYPE_F32, m);
+    power_iteration(pca_params, covariance, result, backend);
+
+    // Free ggml context and backend
+    ggml_free(ctx);
```

Comment:
Remember to free `covariance->data` too (probably alloc the tensor data here before passing it to `compute_covariance`)

Reply:
Instead, allocate the tensor in a CPU buffer and use `ggml_backend_tensor_copy` to copy the result. Or just return the data in a `std::vector`. Don't change `tensor->data` unless you have a very good reason to do that.

Reply:
Got it, I see the issue now! I’ll fix it as well.